Refactor Shared Error UI State

Belal-Alkhatib opened the pull request:
# Pull Request Template

## Description
This PR introduces a centralized and consistent error handling mechanism across the application by creating a shared ErrorUiState and refactoring all relevant ViewModels and UI screens to use it.
The purpose of this change is to improve maintainability, reduce redundancy, and provide a uniform error-handling experience across different features.
It also updates tests to align with the new architecture.
---

## Changes Made
List the changes introduced in this pull request:
### feat: Added error handling to BaseViewModel
- Introduced sealed interface `BaseErrorUiState` (later `ErrorUiState`) to represent:
  - `NoInternetError`
  - `UnknownError`
  - (and others as needed)
- Added `MutableStateFlow` for `errorState`.
- Implemented `updateErrorState` and enhanced `callUseCase` to map `AflamiException` to `ErrorUiState`.

---

### refactor: Migrated multiple screens to use shared `ErrorUiState` instead of local error states
- MyRating
- CategoriesMoviesDetails & CategoriesTvShowsDetails
- Cast
- ActorSearch
- Home
- ListDetails
- Login
- MovieDetails
- SeriesDetails
- TopRated
- Profile, WatchHistory, CountrySearch, UserLists

---

### refactor: Removed obsolete error state classes and properties
- `MyRatingErrorState`
- `CastErrorUiState`
- `ActorSearchErrorState`
- `HomeError`
- `ListDetailsError`
- `networkError` in details screens
- Other unused or redundant error state definitions

---

### refactor: Simplified error handling logic in ViewModels
- Resetting error state on new loads
- Using helper functions like `resetErrorStateToNull` and `updateErrorStateByException`
- Removing redundant `onError` callbacks

---

### UI updates
- Modified affected screens and components to observe `errorState` from ViewModel
- Rendered appropriate error messages or UI based on `ErrorUiState`

---

### test: Updated unit tests
- Refactored tests for all affected ViewModels to validate shared error handling logic


---
### Screenshot: Login Screen as an Example

https://github.com/user-attachments/assets/9ae4b328-477d-4d3a-8fd8-e739ce17c921


## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

Belal-Alkhatib committed:
feature: add error handling to BaseViewModel

Belal-Alkhatib committed:
refactor: use shared error state in MyRating screen

Belal-Alkhatib committed:
refactor: rename `updateErrorState` to `updateErrorStateByException`

Belal-Alkhatib committed:
refactor: remove error state from UI state in categories details

Belal-Alkhatib committed:
refactor: remove CastErrorUiState and use BaseErrorUiState

Belal-Alkhatib committed:
refactor: enhance error handling and state management

Belal-Alkhatib committed:
refactor: simplify error handling in ActorSearchViewModel

Belal-Alkhatib committed:
refactor: simplify error handling in CastViewModel and remove unused error state

Belal-Alkhatib committed:
refactor: update error state type and reset error state

Belal-Alkhatib committed:
refactor: simplify error handling in HomeViewModel and HomeScreen

Belal-Alkhatib committed:
refactor: move network error handling to BaseViewModel

Belal-Alkhatib committed:
refactor: centralize error handling in HomeViewModel

Belal-Alkhatib committed:
refactor: centralize error handling in `ListDetailsViewModel`

Belal-Alkhatib committed:
refactor: simplify LoginViewModel error handling

Belal-Alkhatib committed:
refactor: remove networkError and use ErrorUiState in movie details screen

Belal-Alkhatib committed:
refactor: update series details error UI state

Belal-Alkhatib committed:
refactor: search screens by centralize error handling using base error ui state

Belal-Alkhatib committed:
refactor: update loading and error state management in TopRated screen

Belal-Alkhatib committed:
refactor: centralize error handling and remove redundant error states

This commit introduces a centralized error handling mechanism by:
- Creating a shared `ErrorUiState` to represent different error scenarios (e.g., NoInternetError, UnknownError, LoginErrorState).
- Removing individual error states from ViewModels (Profile, WatchHistory, CountrySearch, UserLists) and their corresponding UI states.
- Updating ViewModels to utilize the shared `ErrorUiState` for error reporting.
- Modifying UI components to consume the centralized `ErrorUiState` for displaying appropriate error messages or UI elements.
- Simplifying error handling logic in ViewModels and UI screens.

This refactoring aims to improve code maintainability, reduce redundancy, and provide a consistent error handling experience across the application.

Belal-Alkhatib committed:
refactor: remove unused exception imports

Belal-Alkhatib committed:
test: migrate view models to use shared error state

Refactored tests for various ViewModels (MovieDetails, SeriesDetails, CountrySearch, ListDetails, ActorSearch, Cast, MyRating, Home, UserLists, Login) to utilize the shared `ErrorUiState` instead of specific error states within each ViewModel's UI state.

This change streamlines error handling and promotes consistency across different parts of the application.

Belal-Alkhatib committed:
Merge branch 'develop' into refactor/shared-error-ui-state

Belal-Alkhatib committed:
refactor: remove unused imports in ViewModels

Belal-Alkhatib committed:
test: fix actor search and list details tests

ThoriaHamdy99 reviewed (changes requested):
Actually, I am not prefer this approach while the Base view model has error UI state that contains all screens error UI states. By this approach all screens will have the state of another screen which does not make sense to access state not my state.

My opinion is --> Make the error UI state in view model has the general error States in all screens like network. And each screen will have another state related to the screen state.

Belal-Alkhatib committed:
Merge branch 'develop' into refactor/shared-error-ui-state

# Conflicts:
#	ui/src/main/java/com/amsterdam/ui/screens/categoriesDetails/movies/CategoriesMoviesDetailsScreen.kt
#	ui/src/main/java/com/amsterdam/ui/screens/lists/ListsScreen.kt
#	ui/src/main/java/com/amsterdam/ui/screens/topRated/TopRatedScreen.kt
#	ui/src/main/java/com/amsterdam/ui/screens/watchHistory/WatchHistoryScreen.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/cast/CastUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/cast/CastViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/categories/CategoriesViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/tvShow/CategoriesTvShowsDetailsUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/home/HomeUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/listDetails/ListDetailsUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/listDetails/ListDetailsViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/lists/ListsUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/lists/WishListsViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/login/LoginUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/login/LoginViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/movieDetails/MovieDetailsViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/myRating/MyRatingViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/profile/ProfileUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/profile/ProfileViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/search/actorSearch/ActorSearchUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/search/countrySearch/CountrySearchViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/search/keywordSearch/SearchViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/seriesDetails/SeriesDetailsViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/topRated/TopRatedViewModel.kt
#	viewModel/src/test/java/com/amsterdam/viewmodel/home/HomeViewModelTest.kt
#	viewModel/src/test/java/com/amsterdam/viewmodel/lists/WishListsViewModelTest.kt

Belal-Alkhatib committed:
Merge branch 'develop' into refactor/shared-error-ui-state

# Conflicts:
#	ui/src/main/java/com/amsterdam/ui/screens/categoriesDetails/movies/CategoriesMoviesDetailsScreen.kt
#	ui/src/main/java/com/amsterdam/ui/screens/categoriesDetails/tvShow/CategoriesTvShowsDetailsScreen.kt
#	ui/src/main/java/com/amsterdam/ui/screens/search/actorSearch/SearchByActorScreen.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/watchHistory/WatchHistoryUiState.kt

Belal-Alkhatib committed:
fix: Remove extra closing brace from the screen.

Belal-Alkhatib committed:
fix: rename GetUserListsUseCase to GetWishListsUseCase

This commit renames `GetUserListsUseCase` to `GetWishListsUseCase` and updates its usage in `WishListsViewModel`. The corresponding UI state mapping function is also renamed from `toUserListItemUiState` to `toWishListItemUiState`.

Belal-Alkhatib committed:
fix: rename GetUserListsUseCase to GetWishListsUseCase

This commit renames `GetUserListsUseCase` to `GetWishListsUseCase` and updates its usage in `WishListsViewModel`. The corresponding UI state mapping function is also renamed from `toUserListItemUiState` to `toWishListItemUiState`.

Belal-Alkhatib committed:
Merge branch 'refs/heads/develop' into refactor/shared-error-ui-state

# Conflicts:
#	ui/src/main/java/com/amsterdam/ui/screens/movieDetails/MovieDetailsScreen.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/movieDetails/MovieDetailsUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/movieDetails/MovieDetailsViewModel.kt

Belal-Alkhatib committed:
refactor: remove network error and unneeded functions

Belal-Alkhatib commented:
> Actually, I am not prefer this approach while the Base view model has error UI state that contains all screens error UI states. By this approach all screens will have the state of another screen which does not make sense to access state not my state.
> 
> My opinion is --> Make the error UI state in view model has the general error States in all screens like network. And each screen will have another state related to the screen state.

Thank you for your feedback @ThoriaHamdy99  🙏 I’ve made another refactor: now generic errors (like network) stay in the base state, while login has its own LoginErrorState inside the LoginUiState class. Please check it out and let me know your opinion 🙂

Belal-Alkhatib committed:
refactor: Decouple login error handling from generic ErrorUiState

This commit introduces the following changes:

-   Creates a new `LoginErrorUiStateMapper.kt` to specifically handle mapping `AflamiException` to `LoginErrorState`.
-   Removes login-specific error states (`InvalidCredentialsError`, `VerificationRequiredError`, `AccountDisabledError`) from `ErrorUiState.kt`.
-   Introduces a dedicated `LoginErrorState` sealed interface within `LoginUiState.kt` to represent login-specific errors.
-   Updates `LoginViewModel.kt` to use the new `LoginErrorState` and `LoginErrorUiStateMapper`.
-   Modifies `LoginScreen.kt` and `GetLoginErrorMessage.kt` to work with the new `LoginErrorState`.
-   Adjusts `ErrorUiStateMapper.kt` to remove the handling of login-specific exceptions.

ahmedmahgoub0 commented on viewModel/src/main/java/com/amsterdam/viewmodel/shared/BaseViewModel.kt:
```diff
@@ -30,9 +32,10 @@ open class BaseViewModel<S, E>(
-
     val state: StateFlow<S> = _state.asStateFlow()
 
+    private val _errorState: MutableStateFlow<ErrorUiState?> = MutableStateFlow(null)
```
why do you need a variable here
you can just send the mapped exception with onError
and each viewmodel take this exception and update its ui state

Belal-Alkhatib replied:
If we just map the exception inside each `ViewModel`, we end up back at the first approach, every `UiState `has to duplicate the same error handling (e.g., mapping `NoInternetError`), which introduces redundancy.

The idea behind the shared `ErrorUiState `is to keep common errors (like network issues) centralized, while still giving flexibility for custom handling through onError. Depending on your case, you can:

- Not call onError → the exception is auto-mapped into the base ErrorUiState.
- Call onError with withAutoUpdateErrorState = false → fully handle the error yourself, no auto-mapping.
- Call onError with withAutoUpdateErrorState = true (default) → auto-mapping + your own custom action.

This way we avoid redundant mappings across view models while still keeping flexibility where needed.

ahmedmahgoub0 replied:
if ui state has error of type ErrorUiState
and onError will send the mapped exception so all what you do is to update the state with mapped exception

ThoriaHamdy99 reviewed (commented):
Nice work

Belal-Alkhatib committed:
Merge branch 'develop' into refactor/shared-error-ui-state

# Conflicts:
#	viewModel/src/main/java/com/amsterdam/viewmodel/continueWatching/ContinueWatchingViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/home/HomeUiState.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/home/HomeViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/seriesDetails/SeriesDetailsViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/topRated/TopRatedViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/watchHistory/WatchHistoryViewModel.kt

Belal-Alkhatib committed:
Refactor: remove unused imports

nabilBouzineDev reviewed (dismissed):
Well done 👏

Belal-Alkhatib committed:
test: remove unused import in WishListsViewModelTest

Belal-Alkhatib committed:
Merge branch 'refs/heads/develop' into refactor/shared-error-ui-state

Belal-Alkhatib committed:
test: Remove redundant advanceUntilIdle() in MovieDetailsViewModelTest

Belal-Alkhatib committed:
Merge branch 'refs/heads/develop' into refactor/shared-error-ui-state

# Conflicts:
#	viewModel/src/main/java/com/amsterdam/viewmodel/lists/WishListsViewModel.kt
#	viewModel/src/main/java/com/amsterdam/viewmodel/topRated/TopRatedViewModel.kt

Belal-Alkhatib committed:
refactor: remove unused imports

Belal-Alkhatib committed:
test: Update error state assertion in TopRatedViewModelTest

Belal-Alkhatib committed:
test: Update errorState assertions in TopRatedViewModelTest

Belal-Alkhatib committed:
test: Use `UnknownError` in top rated view model tests.

Belal-Alkhatib committed:
Merge branch 'refs/heads/develop' into refactor/shared-error-ui-state

# Conflicts:
#	ui/src/main/java/com/amsterdam/ui/screens/myRating/MyRatingScreen.kt

Belal-Alkhatib committed:
refactor: Separate error state in MyRatingScreen
